etc/migration/migrateUsersToShibboleth.js

```diff
@@ -0,0 +1,352 @@
+ */
+function _writeErrorRow(userHash, message) {
+    csvStream.write({
+        'principal_id': userHash.principalId ? userHash.principalId : '',
+        'email': userHash.email ? userHash.email : '',
+        'message': message ? message : ''
+    });
+    errors++;
+}
+
+// Spin up the application container. This will allow us to re-use existing APIs
+OAE.init(config, function(err) {
+    if (err) {
+        log().error({'err': err}, 'Unable to spin up the application server');
+        return _exit(err.code);
+    }
+
+    _getAllUsersForTenant(tenantAlias, function(userHashes) {    
+        _getExternalIds(userHashes, function(usersWithIds) {
+            _mapUsersToShibLogin(usersWithIds, function(err, mappedUsers) {
+                if (err) {
+                    log().error({'err':err}, 'Encountered error when migrating %s users to Shibboleth', tenantAlias);
+                    return _exit(1);
+                }
+
+                if (_.isEmpty(mappedUsers)) {
+                    util.format('No users were migrated for tenant %s', tenantAlias);
+                } else {
+                    util.format('%s users were migrated to Shibboleth logins.', mappedUsers.length);
+                }
+
+                if (errors > 0) {
+                    log().warn('Some users could not be mapped to Shibboleth logins, check the CSV file for more information');
+                }
+
+                return _exit(0);
+            });
+        });
+    });
+});
+
+/**
+ * Page through all the users in the system, and filter them by tenant
+ *
+ * @param  {String}     tenantAlias             The tenant alias to filter by
+ * @param  {Function}   callback                Invoked when users have been collected
+ * @param  {Object[]}   callback.userHashes     An array of user hashes
+ * @api private
+ */
+var _getAllUsersForTenant = function(tenantAlias, callback) {
+    var userHashes = [];
+
+    PrincipalsDAO.iterateAll(['principalId', 'tenantAlias', 'email'], 100, _aggregateUsers, function(err) {
+        if (err) {
+            log().error({'err': err}, 'Failed to iterate all users');
+            return _exit(1);
+        }
+
+        log().info('Found %s users for specified tenant', userHashes.length);
+
+        return callback(userHashes);
+    });
+
+    /*!
+     * Filter users down to those that are part of the specified tenant. Then add them to the `userHashes` array
+     *
+     * @param  {Object[]}   principalHashes     The principals to filter and aggregate
+     * @param  {Function}   callback            Will be invoked when the principals are aggregated
+     */
+    function _aggregateUsers(principalHashes, callback) {
+        log().info('Checking %s principals for tenancy', principalHashes.length);
+        _.chain(principalHashes)
+            .filter(function(principalHash) {
+                return principalHash.tenantAlias === tenantAlias && PrincipalsDAO.isUser(principalHash.principalId) && principalHash.email;
```

Comment:
I presume the script will output these somewhere so we can follow up manually. In reality I think that if we follow the suggestion below to directly parse the email address portion of the google login ids these would be extremely unlikely to occur.